### PR TITLE
Redis implementation

### PR DESCRIPTION
**Pull request by p16i**
In this task, I introduced `async/await` for asynchronous calls to prevent callback stacks. As a results, many blocks of code need to be refactored. From what I see, everything seems to be ok, but it would be better to test a bit more.

REF #4

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,6 +2,8 @@ FROM mhart/alpine-node:6.3
 
 MAINTAINER YouPin Team <dev@youpin.city>
 
+RUN apk add --update g++ make python
```
Comment (parnurzeal) [inline Dockerfile:5]: what is apk?

Comment (supasate) [reply Dockerfile:5]: It's like apt-get but for alpine linux image.

File: docker-compose.yml [hunk 2]
```diff
@@ -5,13 +5,16 @@ services:
     build: .
     depends_on:
       - api
+      - redis
     ports:
       -  "5100:5100"
   api:
-    build: https://github.com/youpin-city/youpin-api.git
+    build: ../youpin-api
```
Comment (supasate) [inline docker-compose.yml:12]: Should we stick to URL? So, it will get the latest image of API and cut off the local dependency.

File: README.md [hunk 3]
```diff
@@ -84,7 +84,7 @@ You should copy the App Secret and Page Access Token. We will use them in our co
 
 10. Go to your created Facebook app. Under `Products` section, `+Add Product` and choose `Webhooks`.
 
-  Enter your `https://<id>.ngrok.io` into `Callback URL`.
+  Enter your `https://<id>.ngrok.io/webhook` into `Callback URL`.
```
Comment (tvirot) [inline README.md:87]: I think we also need another / at the end, right?

File: conversation.js [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
-      this._conversations[userid] = context;
+module.exports = (sessionMaxLength) => {
+  return {
+    sessionPrefix: 'youpin-user:',
+    getContext: async(function(userid) {
+      var context  = JSON.parse(await(redisClient.getAsync( this._buildKey(userid) )))
+      console.log("GET CONTEXT FROM STORE " + JSON.stringify(context) );
+      return context || {};
+    }),
+    updateContext: async(function(userid, context) {
```
Comment (tvirot) [inline conversation.js]: nit: line space between functions?

File: conversation.js [hunk 5]
```diff
@@ -1,26 +1,29 @@
...
-      return {};
-    },
+const redisClient = redis.createClient(redisConf);
 
-    updateContext: function(userid, context) {
-      this._conversations[userid] = context;
+module.exports = (sessionMaxLength) => {
+  return {
+    sessionPrefix: 'youpin-user:',
+    getContext: async(function(userid) {
```
Comment (tvirot) [inline conversation.js]: nit: perhaps a space after `async` and `await` for readability?

**Comment by tvirot** [comment]
Awesome! Overall, this looks good to me 👍

File: docker-compose.yml (hunk 2 above)
Comment (p16i) [reply docker-compose.yml:12]: Does this work with the url? I think in order to run `api`, we need to set up several things such as GCS credentials.

File: README.md (hunk 3 above)
Comment (p16i) [reply README.md:87]: hmm, I'm not sure, but i think either is fine.

File: conversation.js (hunk 4 above)
Comment (p16i) [reply conversation.js]: ok

File: docker-compose.yml (hunk 2 above)
Comment (supasate) [reply docker-compose.yml:12]: You're right. Let's use your change.

**Comment by supasate** [comment]
@tvirot If you run it without any error, you could merge it then.

**Pull request merged**